### PR TITLE
[feat] Consolidate SmartRate functions, add new functions

### PR DESCRIPTION
**Pull request by nwithan8**
# Description

- Add new `SmartRate` service for interacting with the SmartRate API
  - New `EstimateDeliveryDate` function to estimate delivery date based on a list of carriers, to/from ZIP codes and a planned ship date (no existing shipment required).
  - New `RecommendShipDate` function to to recommend ship date based on a list of carriers, to/from ZIP codes and a planned ship date (no existing shipment required).
- New `EstimateDeliveryDate` function in `Shipment` service to estimate delivery date for a shipment based on a planned ship date.
- New `RecommendShipDateForShipment` function in `Shipment` service to recommend ship date for a shipment based on a desired delivery date.
 - New model classes as needed for JSON response to new API functions

# Testing

- Unit tests added, updated accordingly
- Cassettes added, re-recorded accordingly

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

**Review by Justintime50 — COMMENTED** [review]
I think this looks good. As discussed, we'll hold off until this goes GA to merge in and release. I feel splitting SmartRate to its own service is a good idea especially as the product matures and more functionality becomes available. Since everything is static now, it doesn't make sense to keep it in the Shipment service because you have to pass in the shipment ID anyway now, it no longer inherently belongs to Shipment.

File: EasyPost/Services/Beta/SmartRateService.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+using EasyPost.Parameters.SmartRate;
+using EasyPost.Utilities.Internal.Attributes;
+
+namespace EasyPost.Services.Beta
+{
+    /// <summary>
+    ///     Class representing a set of SmartRate-related beta functionality.
+    /// </summary>
+    // ReSharper disable once ClassNeverInstantiated.Global
+    public class SmartRateService : EasyPostService
```
Comment (jchen293) [inline EasyPost/Services/Beta/SmartRateService.cs:16]: Should this class be out of beta when the endpoints are in GA?

File: EasyPost/Services/Beta/SmartRateService.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        /// <returns>A list of <see cref="SmartRate"/>s.</returns>
+        [CrudOperations.Read]
+        public async Task<List<SmartRate>> GetSmartRates(string shipmentId, CancellationToken cancellationToken = default)
+        {
+            return await RequestAsync<List<SmartRate>>(Method.Get, $"shipments/{shipmentId}/smartrate", cancellationToken, rootElement: "result");
+        }
+
+        /// <summary>
+        ///     Retrieve the estimated delivery date of each rate for a <see cref="Shipment"/> via the SmartRates API, based on a specific ship date.
+        ///     <a href="https://www.easypost.com/docs/api#retrieve-estimated-delivery-date-and-total-transit-days-across-all-rates-for-a-shipment">Related API documentation</a>.
```
Comment (jchen293) [inline EasyPost/Services/Beta/SmartRateService.cs:44]: URL seems incorrect to what this function does, you might want to add a TODO and come back to fix it when API docs is added for this endpoint

**Review by jchen293 — COMMENTED** [review]
Code seems good, just two questions. I will approve when the endpoints are in GA

**Comment by Justintime50** [comment]
I'd like this PR to follow the conventions set in the Python one so the experience is consistent across libs, namely the location of the functions in two separate services (keeping the existing functions in the Shipment service), renaming them, and not overloading the functions even though that's a feature of this language (to ensure consistency).

File: EasyPost.Tests/ServicesTests/WithParameters/ShipmentServiceTest.cs [hunk 3]
```diff
@@ -326,6 +327,56 @@ public async Task TestRetrieveEstimatedDeliveryDates()
...
+            List<EstimateDeliveryDateForShipmentResult> ratesWithEstimatedDeliveryDates = await Client.Shipment.EstimateDeliveryDate(shipment.Id, estimateDeliveryDateForShipmentParameters);
+
+            foreach (var rate in ratesWithEstimatedDeliveryDates)
+            {
+                Assert.NotNull(rate.TimeInTransitDetails);
+                Assert.NotNull(rate.TimeInTransitDetails.EasyPostEstimatedDeliveryDate);
+                Assert.NotNull(rate.TimeInTransitDetails.TimeInTransitPercentiles);
+                Assert.NotNull(rate.TimeInTransitDetails.PlannedShipDate);
+            }
+        }
```
Comment (jchen293) [inline EasyPost.Tests/ServicesTests/WithParameters/ShipmentServiceTest.cs:352]: This unit test needs to be removed as we went with the Python lib approach

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,13 @@
 
 ## Next Release
 
+- Add new `SmartRate` service for interacting with the SmartRate API
+  - New `EstimateDeliveryDateForShipment` function to estimate delivery date for a shipment based on a planned ship date.
+  - New `RecommendShipDateForShipment` function to recommend ship date for a shipment based on a desired delivery date.
+  - New `EstimateDeliveryDateForRoute` function to estimate delivery date based on a list of carriers, to/from ZIP codes and a planned ship date (no existing shipment required).
+  - New `RecommendShipDateForRoute` function to to recommend ship date based on a list of carriers, to/from ZIP codes and a planned ship date (no existing shipment required).
+  - New model classes as needed for JSON response to new API functions
+  - Existing SmartRate-related functions and classes in `Shipment` service marked as deprecated and will be removed in a future release
```
Comment (jchen293) [inline CHANGELOG.md:11]: This needs to be updated as we went with the Python lib approach

File: EasyPost/Models/API/EstimateDeliveryDateForZipPairResult.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        ///     The ship date used for the estimates.
+        /// </summary>
+        [JsonProperty("planned_ship_date")]
+        public string? PlannedShipDate { get; set; }
+
+        /// <summary>
+        ///     The estimate results.
+        /// </summary>
+        [JsonProperty("results")]
+        public List<DeliveryDateForZipPairEstimate>? Estimates { get; set; }
```
Comment (jchen293) [inline EasyPost/Models/API/EstimateDeliveryDateForZipPairResult.cs:48]: Our API response is `Results`, we should not change it to `Estimates` in C# exclusively, we should be consistent with other libs. This also apply to other fields (from_zip, to_zip, and carriers_without_tint_estimates)

File: EasyPost/Models/API/TimeInTransitDetailsForShipDate.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    ///     EasyPost's estimated delivery date for the associated <see cref="EstimateDeliveryDateForShipmentResult"/>.
+    /// </summary>
+    [JsonProperty("easypost_estimated_delivery_date")]
+    public DateTime? EasyPostEstimatedDeliveryDate { get; set; }
+
+    /// <summary>
+    ///     Expanded confidence levels for time in transit estimates.
+    /// </summary>
+    [JsonProperty("days_in_transit")]
+    public TimeInTransit? TimeInTransitPercentiles { get; set; }
```
Comment (jchen293) [inline EasyPost/Models/API/TimeInTransitDetailsForShipDate.cs:29]:
```suggestion
    [JsonProperty("days_in_transit")]
    public TimeInTransit? DayInTransit { get; set; }
```

**Review by jchen293 — CHANGES_REQUESTED** [review]

File: EasyPost/BetaClient.cs [hunk 7]
```diff
@@ -40,8 +39,8 @@ public class BetaClient : EasyPostClient
         internal BetaClient(ClientConfiguration configuration)
             : base(configuration)
         {
-            Rate = new RateService(this);
-            ReferralCustomer = new ReferralCustomerService(this);
+            Rate = new Services.Beta.RateService(this);
+            ReferralCustomer = new Services.Beta.ReferralCustomerService(this);
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost/BetaClient.cs:43]: We are not going to deprecate or remove those beta functions in this current PR, that's a separate task.

File: EasyPost/BetaClient.cs [hunk 8]
```diff
@@ -26,12 +25,12 @@ public class BetaClient : EasyPostClient
...
         ///     Access beta Rate-related functionality.
         /// </summary>
-        public RateService Rate { get; }
+        public Services.Beta.RateService Rate { get; }
 
         /// <summary>
         ///     Access beta Referral Customer-related functionality.
         /// </summary>
-        public ReferralCustomerService ReferralCustomer { get; }
+        public Services.Beta.ReferralCustomerService ReferralCustomer { get; }
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost/BetaClient.cs:33]: No, it's just to clean up the import by making the import path more explicit.

File: CHANGELOG.md [hunk 9]
```diff
@@ -2,6 +2,13 @@
 
 ## Next Release
 
+- Add new `SmartRate` service for interacting with the SmartRate API
+    - New `EstimateDeliveryDateForShipment` function to estimate delivery date for a shipment based on a planned ship date.
+    - New `RecommendShipDateForShipment` function to recommend ship date for a shipment based on a desired delivery date.
+    - New `EstimateDeliveryDateForRoute` function to estimate delivery date based on a list of carriers, to/from ZIP codes and a planned ship date (no existing shipment required).
+    - New `RecommendShipDateForRoute` function to to recommend ship date based on a list of carriers, to/from ZIP codes and a planned ship date (no existing shipment required).
+    - New model classes as needed for JSON response to new API functions
+    - Existing SmartRate-related functions and classes in `Shipment` service marked as deprecated and will be removed in a future release
```
Comment (jchen293) [inline CHANGELOG.md:11]: Please update the changelog, the function name doesn't match what we have and we are not deprecating the existing function in `Shipment` service. I will approve once this is addressed

**Review by jchen293 — APPROVED** [review]

**Pull request merged**